e2e/node/request-metadata.test.ts

```diff
@@ -0,0 +1,265 @@
+      await session.logout();
+    });
+
+    describe("authenticated", () => {
+      it("can read back predefined headers set by the client on a successful request", async () => {
+        let responseHeaders: Headers = new Headers();
+        const readHeaders = (headers: Headers) => {
+          responseHeaders = headers;
+        };
+        const spiedFetch = jest.spyOn(session, "fetch");
+        const customFetch = customizeHeaders(
+          session.fetch,
+          requestHeaders,
+          readHeaders,
+        );
+
+        await expect(
+          getSolidDataset(pod, { fetch: customFetch }),
+        ).resolves.not.toThrow();
+
+        expect(spiedFetch).toHaveBeenCalledWith(
+          expect.anything(),
+          expect.objectContaining({
+            headers: expect.objectContaining({
+              [Object.keys(returnedHeaders)[0]]:
+                Object.values(returnedHeaders)[0],
+              "non-returned-header": "some-non-returned-value",
+            }),
+          }),
+        );
+        expect(responseHeaders.get(Object.keys(returnedHeaders)[0])).toContain(
+          Object.values(returnedHeaders)[0],
+        );
+        expect(responseHeaders.get("non-returned-header")).toBeNull();
+      });
+
+      it("can read back predefined headers set by the client on a failed request", async () => {
+        let responseHeaders: Headers = new Headers();
+        const readHeaders = (headers: Headers) => {
+          responseHeaders = headers;
+        };
+        const spiedFetch = jest.spyOn(session, "fetch");
+        const customFetch = customizeHeaders(
+          session.fetch,
+          requestHeaders,
+          readHeaders,
+        );
+
+        // The response will be a 404
+        await expect(() =>
+          getSolidDataset(new URL("non-existing-resource", pod).href, {
+            fetch: customFetch,
+          }),
+        ).rejects.toThrow();
+
+        expect(spiedFetch).toHaveBeenCalledWith(
+          expect.anything(),
+          expect.objectContaining({
+            headers: expect.objectContaining({
+              [Object.keys(returnedHeaders)[0]]:
+                Object.values(returnedHeaders)[0],
+            }),
+          }),
+        );
+        expect(responseHeaders.get(Object.keys(returnedHeaders)[0])).toContain(
+          Object.values(returnedHeaders)[0],
+        );
+        expect(responseHeaders.get("non-returned-header")).toBeNull();
+      });
+    });
+
+    describe("unauthenticated", () => {
+      it("can read back predefined headers set by the client on a successful request", async () => {
+        let responseHeaders: Headers = new Headers();
+        const readHeaders = (headers: Headers) => {
+          responseHeaders = headers;
+        };
+        const spiedFetch = jest.spyOn(global, "fetch");
+        const customFetch = customizeHeaders(
+          fetch,
+          requestHeaders,
+          readHeaders,
+        );
+
+        // The response will be a 401
```

Comment:
if this is a `resolves.not.toThrow` is that actually a 401? this comment seems right but the validator seems wrong.

Reply:
`solid-client` acts as a high-level client, and as such it will throw on 401 or 403.